signer/fulcio/fulcio.go

```diff
@@ -15,133 +15,286 @@
 
-	"github.com/sigstore/fulcio/pkg/api"
+	"github.com/mattn/go-isatty"
+	fulciopb "github.com/sigstore/fulcio/pkg/generated/protobuf"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
 	"github.com/sigstore/sigstore/pkg/oauthflow"
 	"github.com/sigstore/sigstore/pkg/signature"
 	sigo "github.com/sigstore/sigstore/pkg/signature/options"
 	"github.com/testifysec/go-witness/cryptoutil"
+	"github.com/testifysec/go-witness/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"gopkg.in/square/go-jose.v2/jwt"
 )
 
-func Signer(ctx context.Context, funcioURL string, oidcIssuer string, oidcClientID string) (cryptoutil.Signer, error) {
-	fClient, err := newClient(funcioURL)
+func Signer(ctx context.Context, fulcioURL string, oidcIssuer string, oidcClientID string, tokenOption string) (cryptoutil.Signer, error) {
+	// Parse the Fulcio URL to extract its components
+	u, err := url.Parse(fulcioURL)
 	if err != nil {
 		return nil, err
 	}
 
-	key, err := rsa.GenerateKey(rand.Reader, 4096)
-	if err != nil {
-		return nil, err
+	// Get the scheme, default to HTTPS if not present
+	scheme := u.Scheme
+	if scheme == "" {
+		scheme = "https"
+	}
+
+	// Get the port, default to 443
+	port := 443
+	if u.Port() != "" {
+		p, err := strconv.Atoi(u.Port())
+		if err != nil {
+			return nil, fmt.Errorf("invalid port in Fulcio URL: %s", u.Port())
+		}
+		port = p
 	}
 
-	fulcioSigner, err := signature.LoadSigner(key, crypto.SHA256)
+	// Get the host, return an error if not present
+	if u.Host == "" {
+		return nil, errors.New("fulcio URL must include a host")
+	}
+
+	// Make insecure true only if the scheme is HTTP
+	insecure := scheme == "http"
+
+	fClient, err := NewClient(ctx, scheme+"://"+u.Host, port, insecure)
 	if err != nil {
 		return nil, err
 	}
 
-	certResp, err := getCert(fulcioSigner.(*signature.RSAPKCS1v15Signer), fClient, oidcIssuer, oidcClientID)
+	key, err := rsa.GenerateKey(rand.Reader, 4096)
 	if err != nil {
 		return nil, err
 	}
 
-	block, _ := pem.Decode(certResp.CertPEM)
-	if block == nil {
-		return nil, fmt.Errorf("failed to parse certificate PEM")
-	}
+	var raw string
+
+	switch {
+	case tokenOption == "" && os.Getenv("GITHUB_ACTIONS") == "true":
+		tokenURL := os.Getenv("ACTIONS_ID_TOKEN_REQUEST_URL")
+		if tokenURL == "" {
+			return nil, errors.New("ACTIONS_ID_TOKEN_REQUEST_URL is not set")
+		}
+
+		requestToken := os.Getenv("ACTIONS_ID_TOKEN_REQUEST_TOKEN")
+		if requestToken == "" {
+			return nil, errors.New("ACTIONS_ID_TOKEN_REQUEST_TOKEN is not set")
+		}
+
+		raw, err = fetchToken(tokenURL, requestToken, "sigstore")
+		if err != nil {
+			return nil, err
+		}
+
+	case tokenOption != "":
+		raw = tokenOption
 
-	if block.Type != "CERTIFICATE" {
-		return nil, fmt.Errorf("failed to parse certificate PEM")
+	case tokenOption == "" && isatty.IsTerminal(os.Stdin.Fd()):
+		tok, err := oauthflow.OIDConnect(oidcIssuer, oidcClientID, "", "", oauthflow.DefaultIDTokenGetter)
```

Comment:
I think we'll eventually need to find a way to break this out of the library.  I don't think anything that will take over user interaction belongs in library code.  This is fine for now but something to mull over.

Reply:
I'm open to ideas